### PR TITLE
Add `UniqueParameterIds`, `UniqueTypeIds`, `ParameterName`

### PR DESCRIPTION
**Pull request by michal-pekacki**
### Issues addressed by this PR

Closes #

Added method collecting unique parameter names for the collection of the elements.

### CODE REVIEW

File: Revit_Core_Engine/Query/UniqueParameterNames.cs [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+    public static partial class Query
+    {
+        /***************************************************/
+        /****              Public methods               ****/
+        /***************************************************/
+
+        [Description("Returns unique parameter names for the collection of the elements.")]
+        [Input("elements", "Elements to get the parameter names from.")]
+        [Output("paramterNames", "Output containing instance parameter names as Item1 and type paramter names as Item2.")]
+        public static Output<HashSet<string>, HashSet<string>> UniqueParametersNames(this IEnumerable<Element> elements)
```
Comment (pawelbaran) [inline Revit_Core_Engine/Query/UniqueParameterNames.cs:41]:
```suggestion
        public static Output<HashSet<string>, HashSet<string>> UniqueParameterNames(this IEnumerable<Element> elements)
```

File: Revit_Core_Engine/Query/UniqueParameterNames.cs [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+            Dictionary<Document, List<Element>> elementsByDocumentDictionary = elements.ElementsByDocumentDictionary();
+
+            foreach (var elementsInDocPair in elementsByDocumentDictionary)
+            {
+                Document doc = elementsInDocPair.Key;
+                List<Element> elementsFromOneDocument = elementsInDocPair.Value;
+
+                if (!elementsFromOneDocument.Any())
+                    continue;
```
Comment (pawelbaran) [inline Revit_Core_Engine/Query/UniqueParameterNames.cs:60]: This will never occur

File: Revit_Core_Engine/Query/UniqueParameterNames.cs [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+                Dictionary<ElementId, List<Element>> elementsByCategoryDictionary = elementsFromOneDocument.ElementsByCategoryDictionary();
+
+                foreach (var elementsByCat in elementsByCategoryDictionary)
+                {
+                    List<Element> elementsOfCat = elementsByCat.Value;
+
+                    if (!elementsOfCat.Any()) 
+                        continue;
```
Comment (pawelbaran) [inline Revit_Core_Engine/Query/UniqueParameterNames.cs:69]: This will never occur

File: Revit_Core_Engine/Query/UniqueParameterNames.cs [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+                    elementsByDoc.Add(element.Document, new List<Element> { element });
+            }
+
+            return elementsByDoc;
+        }
+
+        /***************************************************/
+
+        [Description("Creates dictionary for elements per category.")]
+        private static Dictionary<ElementId, List<Element>> ElementsByCategoryDictionary(this IEnumerable<Element> elements)
```
Comment (pawelbaran) [inline Revit_Core_Engine/Query/UniqueParameterNames.cs:110]:
```suggestion
        private static Dictionary<ElementId, List<Element>> ElementsByCategory(this IEnumerable<Element> elements)
```

File: Revit_Core_Engine/Query/UniqueParameterNames.cs [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+            Output<HashSet<string>, HashSet<string>> parameterNames = new Output<HashSet<string>, HashSet<string>>
+            {
+                Item1 = new HashSet<string>(),
+                Item2 = new HashSet<string>()
+            };
+
+            if (elements == null || !elements.Any())
+                return parameterNames;
+
+            Dictionary<Document, List<Element>> elementsByDocumentDictionary = elements.ElementsByDocumentDictionary();
```
Comment (pawelbaran) [inline Revit_Core_Engine/Query/UniqueParameterNames.cs:52]:
```suggestion
            Dictionary<Document, List<Element>> elementsByDocument = elements.ElementsByDocument();
```

File: Revit_Core_Engine/Query/UniqueParameterNames.cs [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+
+            foreach (var elementsInDocPair in elementsByDocumentDictionary)
+            {
+                Document doc = elementsInDocPair.Key;
+                List<Element> elementsFromOneDocument = elementsInDocPair.Value;
+
+                if (!elementsFromOneDocument.Any())
+                    continue;
+
+                Dictionary<ElementId, List<Element>> elementsByCategoryDictionary = elementsFromOneDocument.ElementsByCategoryDictionary();
```
Comment (pawelbaran) [inline Revit_Core_Engine/Query/UniqueParameterNames.cs:62]:
```suggestion
                Dictionary<ElementId, List<Element>> elementsByCategory = elementsFromOneDocument.ElementsByCategory();
```

File: Revit_Core_Engine/Query/UniqueParameterNames.cs [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+                    List<Element> elementsOfCat = elementsByCat.Value;
+
+                    if (!elementsOfCat.Any()) 
+                        continue;
+
+                    HashSet<int> instanceParameterIds = elementsOfCat.UniqueParametersIds();
+                    IEnumerable<string> instanceParameterNames = instanceParameterIds.Select(x => doc.ParameterName(x));
+
+                    foreach (string name in instanceParameterNames)
+                        parameterNames.Item1.Add(name);
```
Comment (pawelbaran) [inline Revit_Core_Engine/Query/UniqueParameterNames.cs:75]:
```suggestion
                    parameterNames.Item1.UnionWith(instanceParameterNames);
```

File: Revit_Core_Engine/Query/UniqueParameterNames.cs [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+
+                    IEnumerable<Element> elementTypes = elementsOfCat.UniqueTypeIds().Select(x => doc.GetElement(new ElementId(x)));
+                    HashSet<int> typeParametereIds = elementTypes.UniqueParametersIds();
+                    IEnumerable<string> typeParameterNames = typeParametereIds.Select(x => doc.ParameterName(x));
+
+                    foreach (string name in typeParameterNames)
+                        parameterNames.Item2.Add(name);
```
Comment (pawelbaran) [inline Revit_Core_Engine/Query/UniqueParameterNames.cs:82]:
```suggestion
                    parameterNames.Item2.UnionWith(typeParameterNames);
```

File: Revit_Core_Engine/Query/UniqueParameterNames.cs [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+
+            return parameterNames;
+        }
+
+        /***************************************************/
+
+        [Description("Creates dictionary for elements per document.")]
+        private static Dictionary<Document, List<Element>> ElementsByDocumentDictionary(this IEnumerable<Element> elements)
+        {
+            Dictionary<Document, List<Element>> elementsByDoc = new Dictionary<Document, List<Element>>();
```
Comment (pawelbaran) [inline Revit_Core_Engine/Query/UniqueParameterNames.cs:94]: Probably no effect on performance, but you would encode it in a single line by chaining Linq `GroupBy` and `ToDictionary`

File: Revit_Core_Engine/Query/UniqueParameterNames.cs [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+            }
+
+            return elementsByCat;
+        }
+
+        /***************************************************/
+
+        [Description("Returns unique parameter ids for the collection of the elements.")]
+        private static HashSet<int> UniqueParametersIds(this IEnumerable<Element> elementsFromOneDocument)
+        {
```
Comment (pawelbaran) [inline Revit_Core_Engine/Query/UniqueParameterNames.cs:129]: Here you could have one more hashset that would capture visited families to check element types only once per family:

```C#
HashSet<int> ids = new HashSet<int>();
HashSet<string> visitedFamilies = new HashSet<string>();
foreach (Element element in elementsFromOneDocument)
{
    
    if (element is ElementType)
    {
        string familyName = ((ElementType).element).FamilyName;
        if (visitedFamilies.Contains(familyName)
            continue;
        else
            visitedFamilies.Add(familyName);
    }
    
    foreach (Parameter parameter in element.ParametersMap)
    {
        ids.Add(parameter.Id.IntegerValue);
    }
}
```

**IMPORTANT:** highlight it only works for elements of same category!!

File: Revit_Core_Engine/Query/UniqueParameterNames.cs [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+            HashSet<int> ids = new HashSet<int>();
+            Document doc = elementsFromOneDocument.FirstOrDefault().Document;
+
+            foreach (Element element in elementsFromOneDocument)
+            {
+                Element elementType = doc.GetElement(element.GetTypeId());
+
+                if (elementType != null)
+                    ids.Add(elementType.Id.IntegerValue);
```
Comment (pawelbaran) [inline Revit_Core_Engine/Query/UniqueParameterNames.cs:156]: No need to call `doc.GetElement`? Just cull out -1 in the end

File: Revit_Core_Engine/Query/UniqueParameterNames.cs [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+
+        [Description("Returns name of the paramter based on id and document it belongs.")]
+        private static string ParameterName(this Document doc, int id)
+        {
+            string parameterName;
+
+            if (id < 0)
+                parameterName = LabelUtils.GetLabelFor((BuiltInParameter)id);
+            else
+                parameterName = (doc.GetElement(new ElementId(id)) as ParameterElement).GetDefinition().Name;
```
Comment (pawelbaran) [inline Revit_Core_Engine/Query/UniqueParameterNames.cs:172]: Curious if `ParameterElement` does not already have the name we are looking for?

**Review by pawelbaran — CHANGES_REQUESTED** [review]

File: Revit_Core_Engine/Query/UniqueParameterNames.cs (hunk 10 above)
Comment (michal-pekacki) [reply Revit_Core_Engine/Query/UniqueParameterNames.cs:129]: Added 👍

**Comment by michal-pekacki** [comment]
@BHoMBot check compliance

**Comment by michal-pekacki** [comment]
@BHoMBot check compliance

File: Revit_Core_Engine/Query/ParameterName.cs [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+        [Input("id", "Id of the parameter.")]
+        [Output("parameterName", "Parameter name.")]
+        public static string ParameterName(this Document doc, int id)
+        {
+            string parameterName;
+
+            if (id < 0)
+                parameterName = LabelUtils.GetLabelFor((BuiltInParameter)id);
+            else
+                parameterName = (doc.GetElement(new ElementId(id)) as ParameterElement).Name;
```
Comment (pawelbaran) [inline Revit_Core_Engine/Query/ParameterName.cs:46]:
```suggestion
                parameterName = (doc.GetElement(new ElementId(id)) as ParameterElement)?.Name;
```

**Comment by michal-pekacki** [comment]
@BHoMBot check compliance

File: Revit_Core_Engine/Query/UniqueParametersIds.cs [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+            IEnumerable<Element> elements = new FilteredElementCollector(document).OfCategory(category).WhereElementIsNotElementType();
+
+            return UniqueTypeParametersIds(elements);
+        }
+
+        /***************************************************/
+
+        [Description("Returns unique instance parameter ids for the collection of the elements.")]
+        [Input("elementsFromOneDocument", "Elements from the same document to get the unique parameter ids from.")]
+        [Output("parameterIds", "Unique ids of the instance parameters for the collection of the elements.")]
```
Comment (pawelbaran) [inline Revit_Core_Engine/Query/UniqueParametersIds.cs:66]:
```suggestion
        [Output("parameterIds", "Unique ids of instance parameters for the input collection of elements.")]
```

File: Revit_Core_Engine/Query/UniqueParametersIds.cs [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+            
+            return UniqueInstanceParametersIds(elements);
+        }
+
+        /***************************************************/
+
+        [Description("Returns unique type parameter ids for all elements of category.")]
+        [Input("document", "Document where elements belong.")]
+        [Input("category", "Category of the elements.")]
+        [Output("parameterIds", "Unique type parameters ids for the collection of the elements.")]
```
Comment (pawelbaran) [inline Revit_Core_Engine/Query/UniqueParametersIds.cs:54]:
```suggestion
        [Output("parameterIds", "Unique parameter ids of types that belong to the input document and category.")]
```

File: Revit_Core_Engine/Query/UniqueParametersIds.cs [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+    public static partial class Query
+    {
+        /***************************************************/
+        /****              Public methods               ****/
+        /***************************************************/
+
+        [Description("Returns unique instance parameter ids for all elements of category.")]
+        [Input("document", "Document where elements belong.")]
+        [Input("category", "Category of the elements.")]
+        [Output("parameterIds", "Unique instance parameters ids for the collection of the elements.")]
```
Comment (pawelbaran) [inline Revit_Core_Engine/Query/UniqueParametersIds.cs:41]:
```suggestion
        [Output("parameterIds", "Unique parameter ids of elements that belong to the input document and category.")]
```

**Review by pawelbaran — COMMENTED** [review]
Last comments, besides one suggestion: you have two method names in `UniqueParameterIds` file, which the bot will not like. So I would recommend renaming the file to simply `ParameterIds` - then should be all good 👍

**Comment by michal-pekacki** [comment]
@BHoMBot check compliance

**Review by pawelbaran — APPROVED** [review]
I think it was a longer journey than any of us expected @michal-pekacki 😃 Really happy about the result though, I like the code a lot now, happy to approve, thanks! 💪

**Comment by pawelbaran** [comment]
@BHoMBot check compliance

**Comment by michal-pekacki** [comment]
@BHoMBot check required

**Comment by pawelbaran** [comment]
@BHoMBot check ready-to-merge

**Comment by FraserGreenroyd** [comment]
@BHoMBot check ready-to-merge

**Pull request merged**